### ISSUE TITLE
File stream_rpc.sock is left behind with stream_listen enabled which prevents restart

### ISSUE THREAD

**Issue by juumixx** [Kong/kong#6827]
### Summary

Kong leaves a file `/usr/local/kong/stream_rpc.sock` behind when stream_listen is enabled which prevents restart.

### Steps To Reproduce

1. Setup a streaming port (in my case `stream_listen = 0.0.0.0:8883 backlog=16384 ssl`)
2. Start kong (`systemctl start kong`)
3. Kong generates a file `/usr/local/kong/nginx-kong-stream.conf` which contains in the end:
```
server {        # ignore (and close }, to ignore content)
    listen unix:/usr/local/kong/stream_rpc.sock udp;
    error_log  logs/error.log debug;
    content_by_lua_block {
        Kong.stream_api()
    }
}
```
4. Kong also creates a socket file `/usr/local/kong/stream_rpc.sock`
5. Restart kong
6. Kong fails with error `[emerg] 40111#0: bind() to unix:/usr/local/kong/stream_rpc.sock failed (98: Address already in use)`

Deleting the socket file manually allows me starting kong again, but automatic restart is prevented.

### Additional Details & Logs

- Kong version 2.3.1
- Operating system: Ubuntu 20.04

**Comment by bungle** [comment]
@juumixx I tested this on my mac, but I didn't see same issue. We need to further investigate what is going on.

**Comment by bungle** [comment]
@juumixx, actually now I got it replicated:

```bash
$ KONG_LOG_LEVEL=warn KONG_STREAM_LISTEN="0.0.0.0:8883 backlog=16384 ssl" ./bin/kong quit
Kong stopped (gracefully)
(master) bungle@pro:~/Sources/kong$ KONG_LOG_LEVEL=warn KONG_STREAM_LISTEN="0.0.0.0:8883 backlog=16384 ssl" ./bin/kong start
Error: ./kong/cmd/start.lua:64: nginx: [emerg] bind() to unix:/usr/local/kong/stream_rpc.sock failed (48: Address already in use)
nginx: [emerg] bind() to unix:/usr/local/kong/stream_rpc.sock failed (48: Address already in use)
nginx: [emerg] bind() to unix:/usr/local/kong/stream_rpc.sock failed (48: Address already in use)
nginx: [emerg] bind() to unix:/usr/local/kong/stream_rpc.sock failed (48: Address already in use)
nginx: [emerg] bind() to unix:/usr/local/kong/stream_rpc.sock failed (48: Address already in use)
nginx: [emerg] still could not bind()


  Run with --v (verbose) or --vv (debug) for more details
```

And in logs:

```
2021/02/11 11:34:16 [notice] 83220#0: signal process started
2021/02/11 11:34:23 [emerg] 83400#0: bind() to unix:/usr/local/kong/stream_rpc.sock failed (48: Address already in use)
2021/02/11 11:34:23 [emerg] 83400#0: bind() to unix:/usr/local/kong/stream_rpc.sock failed (48: Address already in use)
2021/02/11 11:34:23 [emerg] 83400#0: bind() to unix:/usr/local/kong/stream_rpc.sock failed (48: Address already in use)
2021/02/11 11:34:23 [emerg] 83400#0: bind() to unix:/usr/local/kong/stream_rpc.sock failed (48: Address already in use)
2021/02/11 11:34:23 [emerg] 83400#0: bind() to unix:/usr/local/kong/stream_rpc.sock failed (48: Address already in use)
2021/02/11 11:34:23 [emerg] 83400#0: still could not bind()
```

**Comment by rainest** [comment]
Weird. Locally I can only reproduce this with `kong quit` (SIGQUIT), same as Aapo. `kong stop` (SIGTERM) does remove the socket file. Our Helm chart helper (for checking migration completeness) is using `kong stop` though; unclear why it's affected.

**Comment by gszr** [comment]
This is fixed in 2.4. Closing the issue.

**Comment by andrew-landsverk-win** [comment]
I believe I'm still seeing this issue in Kong 2.4 with `stream_config.sock` in Kubernetes. My test was to "pull the plug" on the whole cluster, upon cluster restart, my kong pods do not start until re-created manually (We're using the Kong Helm chart).

`
2021/06/29 14:09:02 [emerg] 1#0: bind() to unix:/kong_prefix/stream_config.sock failed (98: Address in use)
nginx: [emerg] bind() to unix:/kong_prefix/stream_config.sock failed (98: Address in use)
`

Ideally, we wouldn't have any manual intervention in the case of a full cluster shutdown (improperly or not). Please let me know how I can assist.

Thanks!

Kong version: `image: kong:2.4`

**Comment by andrew-landsverk-win** [comment]
Just a quick follow up. It also happens if a single node gets shut down improperly.

**Comment by morningspace** [comment]
Is this issue resolved in latest Kong version? I are seeing the same issue in Kubernetes cluster.

**Comment by Rabia-Riaz-SE** [comment]
I have encounter same issue for ungraceful shutdown even with kong version 3.1.1

**Comment by gszr** [comment]
@Rabia-Riaz-SE, If you are seeing this behavior in a consistent way, I encourage you to open a new issue, making sure to include precise reproduction steps, as described in the [issue template](https://github.com/Kong/kong/issues/new?assignees=&labels=&template=bug_report.yaml). It's very important that you fill in all fields in the form. Thanks in advance!
